Add BearSSL documentation

earlephilhower opened the pull request:
Document the BearSSL::WiFiClientSecure, ::WiFiServerSecure, and the
helper classes required to use them.

Fixes #1743 and other SSL documentation issues.

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+
+TLS supports the notion of a session (completely independent and different from HTTP sessions) which allow clients to reconnect to a server without having to renegotiate encryption settings or validate X509 certificates.  This can save significant time (3-4 seconds in the case of EC keys) and can help save power by allowing the ESP8266 to sleep for a long time, reconnect and transmit some samples ...
+
+`BearSSLSession` is an opaque class.  Use the `BearSSL::WiFiClientSecure.setSession(&BearSSLSession)` method to apply it before the first `BearSSL::WiFiClientSecure.connect()` and it will be updated with session parameters during the operation of the connection.  After the connection had had `.close()` called on it, serialize the `BearSSLSession` object to stable store (EEPROM, RTC RAM, etc) and  ...
```
stable store => stable storage?

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+
+X509 certificates are used to identify peers in TLS connections.  Normally only the server identifies itself, but the client can also supply an X509 certificate if desired (this is often done in MQTT applications).  The certificate contains many fields, but the most interesting in our applications are the name, the public key, and potentially a chain of signing that leads back to a trusted author ...
+
+Any call that takes an X509 certificate can also take a list of X509 certificates, so there is no special `X509` class, simply `BearSSLX509List` (which may only contain a sincle certificate).
```
sincle => single

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+    ...or...
+    BearSSLX509List(const uint8_t *derCert, size_t derLen);
+
+If you need to add additional certificates (unlikely in normal operation, the `::append()` operation can be used.
```
missing closing ), maybe before the comma?

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+setFingerprint(const uint8_t fp[20]) / setFingerprint(const char *fpStr)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Verify the SHA1 fingerprint of the certificate returned matches this one.  If the server certificate changes, it will fail.  If an array of 20 bytes are sent in, it is assumed they are the binary SHA1 values.  If a `char*` string is passed in, it is parsed as a series of human-readable hex values separated by spaces or colons (i.e. `setFingerprint("00:01:02:03:...:1f");`)
```
i.e. => e.g.

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+Client Certificates (Proving I'm who I say I am to the server)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+TLS servers can request that a client identify themselves with an X509 certificate signed by a trust anchor it honors (i.e. a global TA or a private CA).  This is commonly done for applicaitons like MQTT.  By default the client doesn't sent a certificate, and in cases where a certificate is required the server will disconnect and no connection will be possible.
```
sent => send

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+setClientRSACert / setClientECCert
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Sets a client certificate to send to a TLS server that requests one.  It should be called before `connect()` to add a certificate to the client in case the server requests is.  Note that certificates include both a certificate and a private key.  Both should be provided to you by your certificate generator.  Elliptic Curve (EC) keys require additional information, as shown in the prototype.
```
requests is => requests it

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+
+Once you have verified (or know beforehand) that MFLN is supported you can use this call to set the size of memory buffers allocated by the connection object.  This must be called **before** `connect()` or it will be ignored.
+
+In certain applications where the TLS server does not support MFLN (not many do as of this writing as it is relatively new to OpenSSL) but you control both the ESP8266 and the server to which it is communicating, you may still be able to `setBufferSizes()` smaller if you guarantee no chunk of data will overflow those buffers.
```
OpenSSL) but you => OpenSSL), but you (add comma)

devyte commented on doc/esp8266wifi/bearssl-client-secure-class.rst:
```diff
@@ -0,0 +1,213 @@
+
+BearSSL can fail in many more unique and interesting ways then the deprecated axTLS.  Use these calls to get more information when something fails.  
+
+getLastSSLError(char *dest = NULL, size_t len = 0)
```
Is the user supposed to allocate some buffer?
Or is the arg supposed to be a pointer to the base of a static-sized buffer?
Or is the pointer set to an internal static string? If so, should the args be const?

devyte commented on doc/esp8266wifi/bearssl-server-secure-class.rst:
```diff
@@ -0,0 +1,44 @@
+setBufferSizes(int recv, int xmit)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Similar to the `BearSSL::WiFiClientSecure` methos, sets the receive and transmit buffer sizes.  Note that servers cannot request a buffer size from the client, so if these are shrunk and the client tries to send a chunk larger than the receive buffer, it will always fail.  This must be called before the server is 
```
methos => method

earlephilhower commented:
Thanks, @devyte, for the read-through.  Updated with your grammar and spelling (plus MSFT Word spellcheck) and cleaned up the `getLastError()` bit to be more specific that user allocates and manages the string buffer optionally passed in.
